tf2onnx/onnx_opset/tensor.py

```diff
@@ -637,6 +637,161 @@ def version_4(cls, ctx, node, **kwargs):
+                begin = ctx.make_const(
+                    utils.make_name("begin_masked"),
+                    begin.get_tensor_value(as_list=False) * new_begin_mask
+                )
+            else:
+                begin_mask_const = ctx.make_const(
+                    utils.make_name("begin_mask"),
+                    new_begin_mask
+                )
+                begin = ctx.make_node(
+                    "Mul", [begin.output[0], begin_mask_const.output[0]],
+                    op_name_scope=node.name
+                )
+        # mask end
+        new_end_mask = np.array(new_end_mask, dtype=np_dtype)
+        end_output = end.output[0]
+        if not np.all(new_end_mask == min_size):
+            if end.is_const():
+                end = ctx.make_const(
+                    utils.make_name("end_masked"),
+                    np.maximum(end.get_tensor_value(as_list=False), new_end_mask)
+                )
+                end_output = end.output[0]
+            else:
+                end_mask_const = ctx.make_const(
+                    utils.make_name("end_mask"),
+                    np.array(new_end_mask, dtype=np_dtype)
+                )
+                end_output = utils.make_name("{}__end".format(node.name))
+                math.make_minmax_op(ctx, "Max", [end.output[0], end_mask_const.output[0]], [end_output])
+        # mask strides for shrink
+        shrink_strided_mask = np.array(shrink_strided_mask, dtype=np_dtype)
+        strides_output = strides.output[0]
+        if not np.all(shrink_strided_mask == min_size):
+            if strides.is_const():
+                strides = ctx.make_const(
+                    utils.make_name("strides_masked"),
+                    np.maximum(strides.get_tensor_value(as_list=False), shrink_strided_mask)
+                )
+                strides_output = strides.output[0]
+            else:
+                shrink_strided_mask_const = ctx.make_const(
+                    utils.make_name("strides_mask"),
+                    np.array(shrink_strided_mask, dtype=np_dtype)
+                )
+                strides_output = utils.make_name("{}__strides".format(node.name))
+                math.make_minmax_op(
+                    ctx, "Max",
+                    [strides.output[0], shrink_strided_mask_const.output[0]],
+                    [strides_output]
+                )
+        # create axes input
+        axes_const = ctx.make_const(
+            utils.make_name("slice_axes"),
+            np.arange(input_rank, dtype=np_dtype)
+        )
+        axes_output = axes_const.output[0]
+
+        inputs_map = {
+            "data": node.input[0],
+            "starts": begin.output[0],
+            "ends": end_output,
+            "steps": strides_output,
+            "axes": axes_output
+        }
+        kwargs = {**inputs_map, "outputs": node.output}
+        node = GraphBuilder(ctx).make_slice(kwargs, name=node.name, dtypes=out_dtypes, shapes=out_shapes)
```

Comment:
if needs_squeeze is true, are the out_shapes correct here?

Reply:
of course, we always respect tf shape.